src/fusion.cpp

```diff
@@ -810,562 +378,79 @@ miopenStatus_t BiasFusionOpDescriptor::SetArgs(OperatorArgs& args,
+        return miopenStatusUnsupportedOp;
     else
     {
-        auto default_args = lu.GetKernelArgs(handle);
-        if(default_args.empty())
-        {
-            MIOPEN_THROW(miopenStatusInternalError,
-                         "Default kernel args no supplied in metadata graph");
-        }
-        for(auto& arg : default_args)
+        for(const auto& sol : sols)
         {
-            MIOPEN_LOG_I2("Setting arg: " + arg.key);
-            switch(arg.type)
-            {
-            case OpArg:
-                if(arg.op_idx < op_map.size())
-                {
-                    auto& op = op_map.at(arg.op_idx);
-                    auto k   = op->GetArgKey(arg.key);
-                    arg_keys.emplace_back(
-                        k, arg.default_val.is_ptr ? Pointer : Scalar, arg.default_val.size());
-                    break;
-                }
-                else
-                {
-                    arg_keys.emplace_back(
-                        arg.key, Default, arg.default_val.size(), arg.default_val);
-                }
-                break;
-            case OpAttr:
-                if(arg.op_idx < op_map.size())
-                {
-                    auto& op     = op_map.at(arg.op_idx);
-                    auto op_attr = op->GetOpAttr(arg.key);
-                    arg_keys.emplace_back(arg.key, Default, op_attr.size(), op_attr);
-                }
-                else
-                {
-                    arg_keys.emplace_back(
-                        arg.key, Default, arg.default_val.size(), arg.default_val);
-                }
-                break;
-            case Other:
-                // The operator does not exist in the fusion plan, load the default value
-                arg_keys.emplace_back(arg.key, Default, arg.default_val.size(), arg.default_val);
-                break;
-            case InputTensor:
-                arg_keys.emplace_back("reserved_input_tensor_ptr", Input_Ptr, sizeof(ConstData_t));
-                break;
-            case OutputTensor:
-                arg_keys.emplace_back(
-                    "reserved_output_tensor_ptr", Output_Ptr, sizeof(ConstData_t));
-                break;
-            case DevAttribute: {
-                auto dev_attr = GetDevAttribute(arg.key, handle);
-                arg_keys.emplace_back(arg.key, Default, dev_attr.size(), dev_attr);
-            }
-            break;
-            case InputTensorDesc:
-            case OutputTensorDesc:
-                auto tensor_arg = GetTensorAttr(arg.key);
-                arg_keys.emplace_back(arg.key, Default, tensor_arg.size(), tensor_arg);
-                break;
-            }
+            if(!sol.invoker_factory)
+                MIOPEN_THROW(miopenStatusInternalError,
+                             "Invoker missing from solver " + sol.solver_id);
+            const auto invoker = handle.PrepareInvoker(
+                *sol.invoker_factory, sol.construction_params); // force compilation
```

Comment:
I am not sure I fully understand. The call to `PrepareInvoker` includes a call to `AddKernel` which can force the creation of a `Program` object. Therefore, I was under the impression that the `AddKernel` logic you are referring to is not necessary. Can you please elaborate what am I missing here ?

Reply:
`AddKernel` does not necessarily compile anything: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/bc74d11db5187873204b623eb2d541d40a501745/src/kernel_cache.cpp#L130
So thing may call compilation, but it may not as well and does not "force" it.
And by checking if anything is pre-existing I was talking of adding something like this:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/bc74d11db5187873204b623eb2d541d40a501745/src/include/miopen/find_solution.hpp#L350
at the start of the method to prevent duplicate find and re-creation of invokers if compile has been called twice unless that is intended and wished behaviour.

Reply:
@DrizztDoUrden @JehandadKhan 

👍 Yeah, we need to check if Invoker already exists before creating it.

Some clarifications. This is true:
> The call to PrepareInvoker includes a call to AddKernel... 

This is not:
> which can force the creation of a Program object

Creation of a Program object happens only if the same Program object does not exist. If it exists, it's simply reused.

Otherwise the library creates it by calling `handle.LoadProgram(...)`. The latter method may reuse a code object from the binary cache. 

That is, compilation happens only if:
- Program object does not exist (not registered in the program_map)
- Code object does not reside in the binary cache.

I recommend removing the `// force compilation` comment.

Reply:
Removed